pkg/storage/s3.go

```diff
@@ -404,59 +409,112 @@ func (rs *S3Storage) WalkDir(ctx context.Context, opt *WalkOption, fn func(strin
 		Key:    aws.String(rs.options.Prefix + path),
 	}
 
+	// always set rangeOffset to fetch file size info
+	// s3 endOffset is inclusive
 	var rangeOffset *string
-	if startOffset > 0 {
-		if endOffset > startOffset {
-			rangeOffset = aws.String(fmt.Sprintf("bytes=%d-%d", startOffset, endOffset))
-		} else {
-			rangeOffset = aws.String(fmt.Sprintf("bytes=%d-", startOffset))
-		}
-		input.Range = rangeOffset
+	if endOffset > startOffset {
+		rangeOffset = aws.String(fmt.Sprintf("bytes=%d-%d", startOffset, endOffset-1))
+	} else {
+		rangeOffset = aws.String(fmt.Sprintf("bytes=%d-", startOffset))
 	}
-
+	input.Range = rangeOffset
 	result, err := rs.svc.GetObjectWithContext(ctx, input)
 	if err != nil {
-		return nil, err
+		return nil, rangeInfo{}, err
 	}
 
-	// FIXME: we test in minio, when request with Range, the result.AcceptRanges is a bare string 'range',
-	//  not sure whether this is a feature or bug
-	//if rangeOffset != nil && (result.AcceptRanges == nil || *result.AcceptRanges != *rangeOffset) {
-	//	return nil, errors.Errorf("open file '%s' failed, expected range: %s, got: %v",
-	//		name, *rangeOffset, result.AcceptRanges)
-	//}
+	r, err := parseRangeInfo(result.ContentRange)
+	if err != nil {
+		return nil, rangeInfo{}, errors.Trace(err)
+	}
+
+	if startOffset != r.start || (endOffset != 0 && endOffset != r.end+1) {
+		return nil, r, errors.Errorf("open file '%s' failed, expected range: %s, got: %v",
+			path, *rangeOffset, result.ContentRange)
+	}
 
-	return result.Body, nil
+	return result.Body, r, nil
+}
+
+var (
+	contentRangeRegex = regexp.MustCompile(`bytes (\d+)-(\d+)/(\d+)$`)
+)
+
+func parseRangeInfo(info *string) (rangeInfo, error) {
+	if info == nil || len(*info) == 0 {
+		return rangeInfo{}, errors.New("ContentRange is empty")
+	}
+	subMatches := contentRangeRegex.FindStringSubmatch(*info)
+	if len(subMatches) != 4 {
+		return rangeInfo{}, errors.Errorf("invalid content range: '%s'", *info)
+	}
+
+	start, err := strconv.ParseInt(subMatches[1], 10, 64)
+	if err != nil {
+		return rangeInfo{}, errors.Annotatef(err,
+			"invalid start offset value '%s' in ContentRange '%s'", subMatches[1], *info)
+	}
+	end, err := strconv.ParseInt(subMatches[2], 10, 64)
+	if err != nil {
+		return rangeInfo{}, errors.Annotatef(err,
+			"invalid end offset value '%s' in ContentRange '%s'", subMatches[2], *info)
+	}
+	size, err := strconv.ParseInt(subMatches[3], 10, 64)
+	if err != nil {
+		return rangeInfo{}, errors.Annotatef(err,
+			"invalid size size value '%s' in ContentRange '%s'", subMatches[3], *info)
+	}
+	return rangeInfo{start: start, end: end, size: size}, nil
 }
 
 // s3ObjectReader wrap GetObjectOutput.Body and add the `Seek` method.
 type s3ObjectReader struct {
-	storage *S3Storage
-	name    string
-	reader  io.ReadCloser
-	pos     int64
+	storage   *S3Storage
+	name      string
+	reader    io.ReadCloser
+	pos       int64
+	rangeInfo rangeInfo
+	// reader context used for seek
+	ctx context.Context
```

Comment:
please add a comment explaining why `io.Seeker` must be implemented then.

Reply:
@glorv and i mean mention the that Parquet interface.

Reply:
I think implement `io.Seeker` is not only for read parquet files in lightning, but also should be convenient for other usage in the future. Put the `xitongsys/parquet-go` interface here seems somewhat weird to me

Reply:
but Parquet is currently the only reason we must use the Seeker interface

Reply:
ok, I added a comment for this